File: samples/python/enterprise-agent-tutorial/1-idea-to-prototype/evaluate.py
# ------------------------------------
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.
# ------------------------------------
"""
DESCRIPTION:
    This sample demonstrates how to evaluate the Modern Workplace Assistant
    using the cloud evaluation API with built-in evaluators.

USAGE:
    python evaluate.py

    Before running:
    pip install azure-ai-projects==2.0.0b3 python-dotenv openai

    Set these environment variables:
    1) PROJECT_ENDPOINT - Your Foundry project endpoint
    2) MODEL_DEPLOYMENT_NAME - Model deployment name (e.g., gpt-4o-mini)
"""

# <imports_and_includes>
import os
import time
from typing import Union
from pprint import pprint
from dotenv import load_dotenv
from azure.identity import DefaultAzureCredential
from azure.ai.projects import AIProjectClient
from azure.ai.projects.models import PromptAgentDefinition
from openai.types.eval_create_params import DataSourceConfigCustom
from openai.types.evals.run_create_response import RunCreateResponse
from openai.types.evals.run_retrieve_response import RunRetrieveResponse
# </imports_and_includes>

# <configure_evaluation>
load_dotenv()
endpoint = os.environ["PROJECT_ENDPOINT"]
model_deployment_name = os.environ.get("MODEL_DEPLOYMENT_NAME", "gpt-4o-mini")

with (
    DefaultAzureCredential() as credential,
    AIProjectClient(endpoint=endpoint, credential=credential) as project_client,
    project_client.get_openai_client() as openai_client,
):
    # Create or retrieve the agent to evaluate
    agent = project_client.agents.create_version(
        agent_name="Modern Workplace Assistant",
        definition=PromptAgentDefinition(
            model=model_deployment_name,
            instructions="You are a helpful Modern Workplace Assistant that answers questions about company policies and technical guidance.",
        ),
    )
    print(f"Agent created (id: {agent.id}, name: {agent.name}, version: {agent.version})")

    # Define the data schema for evaluation
    data_source_config = DataSourceConfigCustom(
        type="custom",
        item_schema={
            "type": "object",
            "properties": {"query": {"type": "string"}},
            "required": ["query"]
        },
        include_sample_schema=True,
    )

    # Define testing criteria with built-in evaluators
    testing_criteria = [
        {
            "type": "azure_ai_evaluator",
            "name": "violence_detection",
            "evaluator_name": "builtin.violence",
            "data_mapping": {"query": "{{item.query}}", "response": "{{sample.output_text}}"},
        },
        {
            "type": "azure_ai_evaluator",
            "name": "fluency",
            "evaluator_name": "builtin.fluency",
            "initialization_parameters": {"deployment_name": f"{model_deployment_name}"},
            "data_mapping": {"query": "{{item.query}}", "response": "{{sample.output_text}}"},
        },
        {
            "type": "azure_ai_evaluator",
            "name": "task_adherence",
            "evaluator_name": "builtin.task_adherence",
            "initialization_parameters": {"deployment_name": f"{model_deployment_name}"},
            "data_mapping": {"query": "{{item.query}}", "response": "{{sample.output_items}}"},
        },
    ]

    # Create the evaluation object
    eval_object = openai_client.evals.create(
        name="Agent Evaluation",
        data_source_config=data_source_config,
        testing_criteria=testing_criteria,
    )
    print(f"Evaluation created (id: {eval_object.id}, name: {eval_object.name})")
# </configure_evaluation>

# <run_cloud_evaluation>
    # Define the data source for the evaluation run
    data_source = {
        "type": "azure_ai_target_completions",
        "source": {
            "type": "file_content",
            "content": [
                {"item": {"query": "What is Contoso's remote work policy?"}},
                {"item": {"query": "What are the security requirements for remote employees?"}},
                {"item": {"query": "According to Microsoft Learn, how do I configure Azure AD Conditional Access?"}},
                {"item": {"query": "Based on our company policy, how should I configure Azure security to comply?"}},
            ],
        },
        "input_messages": {
            "type": "template",
            "template": [
                {"type": "message", "role": "user", "content": {"type": "input_text", "text": "{{item.query}}"}}
            ],
        },
        "target": {
            "type": "azure_ai_agent",
            "name": agent.name,
            "version": agent.version,
        },
    }

    # Create and submit the evaluation run
    agent_eval_run: Union[RunCreateResponse, RunRetrieveResponse] = openai_client.evals.runs.create(
        eval_id=eval_object.id,
        name=f"Evaluation Run for Agent {agent.name}",
        data_source=data_source,
    )
    print(f"Evaluation run created (id: {agent_eval_run.id})")
# </run_cloud_evaluation>

# <retrieve_evaluation_results>
    # Poll until the evaluation run completes
    while agent_eval_run.status not in ["completed", "failed"]:
        agent_eval_run = openai_client.evals.runs.retrieve(
            run_id=agent_eval_run.id,
            eval_id=eval_object.id
        )
        print(f"Waiting for eval run to complete... current status: {agent_eval_run.status}")
        time.sleep(5)

    if agent_eval_run.status == "completed":
        print("\n✓ Evaluation run completed successfully!")
        print(f"Result Counts: {agent_eval_run.result_counts}")

        # Retrieve detailed output items
        output_items = list(
            openai_client.evals.runs.output_items.list(
                run_id=agent_eval_run.id,
                eval_id=eval_object.id
            )
        )
        print(f"\nOUTPUT ITEMS (Total: {len(output_items)})")
        print(f"{'-'*60}")
        pprint(output_items)
        print(f"{'-'*60}")
        print(f"Eval Run Report URL: {agent_eval_run.report_url}")
    else:
        print("\n✗ Evaluation run failed.")

    # Cleanup
    openai_client.evals.delete(eval_id=eval_object.id)
    print("Evaluation deleted")

    project_client.agents.delete(agent_name=agent.name)
    print("Agent deleted")
# </retrieve_evaluation_results>

File: samples/typescript/enterprise-agent-tutorial/1-idea-to-prototype/src/evaluate.ts
#!/usr/bin/env node
/**
 * Evaluation Script for Modern Workplace Assistant
 * Tests the agent with predefined business scenarios to assess quality.
 *
 * Updated for Azure AI Agents SDK v2.
 */

// <imports_and_includes>
import * as fs from "fs";
import * as path from "path";
import { chatWithAssistant } from "./main";
import { AIProjectClient } from "@azure/ai-projects";
import { DefaultAzureCredential } from "@azure/identity";
import { config } from "dotenv";
// </imports_and_includes>

config();

const credential = new DefaultAzureCredential();
const project = new AIProjectClient(
  process.env.PROJECT_ENDPOINT || "",
  credential
);

interface Question {
  question: string;
  test_type: string;
  expected_source: string;
  validation: string;
  explanation: string;
}

interface EvaluationResult {
  question: string;
  response: string;
  status: string;
  passed: boolean;
  validation_details: string;
  test_type: string;
  expected_source: string;
  explanation: string;
}

interface Stats {
  passed: number;
  total: number;
}

// <load_test_data>
function loadTestQuestions(filepath: string = "questions.jsonl"): Question[] {
  /**
   * Load test questions from JSONL file
   */
  const questions: Question[] = [];
  const content = fs.readFileSync(filepath, "utf-8");
  const lines = content.split("\n").filter((line) => line.trim());

  for (const line of lines) {
    questions.push(JSON.parse(line));
  }

  return questions;
}
// </load_test_data>

// <validation_functions>
function validateResponse(
  response: string,
  validationType: string,
  expectedSource: string
): { passed: boolean; details: string } {
  /**
   * Validate that the response used the expected tools.
   *
   * Args:
   *   response: The agent's response text
   *   validationType: Type of validation to perform
   *   expectedSource: Expected data source (sharepoint, mcp, or both)
   *
   * Returns:
   *   Object with passed boolean and details string
   */
  const responseLower = response.toLowerCase();

  // Check for Contoso-specific content (indicates SharePoint usage)
  const contosoIndicators = [
    "contoso",
    "90-day probationary period",
    "2-hour incident reporting",
    "company policies",
    "our policy",
    "our remote work policy",
    "our security guidelines",
    "our collaboration standards",
    "our data governance",
  ];
  const hasContosoContent = contosoIndicators.some((indicator) =>
    responseLower.includes(indicator)
  );

  // Check for Microsoft Learn links (indicates MCP usage)
  const learnIndicators = [
    "learn.microsoft.com",
    "docs.microsoft.com",
    "microsoft learn",
    "official documentation",
    "reference link",
    "documentation link",
  ];
  const hasLearnLinks = learnIndicators.some((indicator) =>
    responseLower.includes(indicator)
  );

  // Validate based on expected source
  let passed: boolean;
  let details: string;

  if (expectedSource === "sharepoint") {
    passed = hasContosoContent;
    details = `Contoso-specific content: ${hasContosoContent}`;
  } else if (expectedSource === "mcp") {
    passed = hasLearnLinks;
    details = `Microsoft Learn links: ${hasLearnLinks}`;
  } else if (expectedSource === "both") {
    passed = hasContosoContent && hasLearnLinks;
    details = `Contoso content: ${hasContosoContent}, Learn links: ${hasLearnLinks}`;
  } else {
    passed = false;
    details = "Unknown validation type";
  }

  return { passed, details };
}
// </validation_functions>

// <run_batch_evaluation>
async function runEvaluation(agentName: string): Promise<EvaluationResult[]> {
  /**
   * Run evaluation with test questions.
   *
   * Args:
   *   agentName: The name of the agent to evaluate
   *
   * Returns:
   *   Array of evaluation results for each question
   */
  const questions = loadTestQuestions();
  const results: EvaluationResult[] = [];

  console.log(`🧪 Running evaluation with ${questions.length} test questions...`);
  console.log("=".repeat(70));

  // Track results by test type
  const stats: Record<string, Stats> = {
    sharepoint_only: { passed: 0, total: 0 },
    mcp_only: { passed: 0, total: 0 },
    hybrid: { passed: 0, total: 0 },
  };

  for (let i = 0; i < questions.length; i++) {
    const q = questions[i];
    const testType = q.test_type || "unknown";
    const expectedSource = q.expected_source || "unknown";
    const validationType = q.validation || "default";

    console.log(`\n📝 Question ${i + 1}/${questions.length} [${testType.toUpperCase()}]`);
    console.log(`   ${q.question.substring(0, 80)}...`);

    const { response, status } = await chatWithAssistant(agentName, q.question);

    // Validate response using source-specific checks
    const { passed, details } = validateResponse(
      response,
      validationType,
      expectedSource
    );

    const result: EvaluationResult = {
      question: q.question,
      response: response,
      status: status,
      passed: passed,
      validation_details: details,
      test_type: testType,
      expected_source: expectedSource,
      explanation: q.explanation || "",
    };
    results.push(result);

    // Update stats
    if (stats[testType]) {
      stats[testType].total += 1;
      if (passed) {
        stats[testType].passed += 1;
      }
    }

    const statusIcon = passed ? "✅" : "⚠️";
    console.log(`${statusIcon} Status: ${status} | Tool check: ${details}`);
  }

  console.log("\n" + "=".repeat(70));
  console.log("📊 EVALUATION SUMMARY BY TEST TYPE:");
  console.log("=".repeat(70));

  for (const [testType, data] of Object.entries(stats)) {
    if (data.total > 0) {
      const passRate = (data.passed / data.total) * 100;
      const icon = passRate >= 75 ? "✅" : passRate >= 50 ? "⚠️" : "❌";
      console.log(
        `${icon} ${testType.toUpperCase()}: ${data.passed}/${data.total} passed (${passRate.toFixed(1)}%)`
      );
    }
  }

  return results;
}
// </run_batch_evaluation>

// <evaluation_results>
function calculateAndSaveResults(results: EvaluationResult[]): void {
  /**
   * Calculate pass rate and save results
   */
  const passed = results.filter((r) => r.passed).length;
  const total = results.length;
  const passRate = total > 0 ? (passed / total) * 100 : 0;

  console.log(
    `\n📊 Overall Evaluation Results: ${passed}/${total} questions passed (${passRate.toFixed(1)}%)`
  );

  // Save results
  fs.writeFileSync(
    "evaluation_results.json",
    JSON.stringify(results, null, 2)
  );

  console.log("💾 Results saved to evaluation_results.json");

  // Print summary of failures
  const failures = results.filter((r) => !r.passed);
  if (failures.length > 0) {
    console.log(`\n⚠️  Failed Questions (${failures.length}):`);
    for (const r of failures) {
      console.log(`   - [${r.test_type}] ${r.question.substring(0, 60)}...`);
      console.log(`     Reason: ${r.validation_details}`);
    }
  }
}
// </evaluation_results>

async function createWorkplaceAssistant() {
  /**
   * Create agent - simplified version for evaluation
   */
  console.log("🤖 Creating Modern Workplace Assistant for evaluation...");

  const instructions = `You are a Modern Workplace Assistant specializing in Azure and Microsoft 365 guidance.
Provide comprehensive technical guidance with step-by-step implementation instructions.`;

  const agent = await project.agents.createVersion(
    "modern-workplace-assistant-eval",
    {
      kind: "prompt",
      model: process.env.MODEL_DEPLOYMENT_NAME || "gpt-4o",
      instructions: instructions,
    }
  );

  console.log(`✅ Agent created: ${agent.id}`);
  return agent;
}

async function main(): Promise<void> {
  /**
   * Run evaluation on the workplace assistant using Agent SDK v2.
   */
  console.log("🧪 Modern Workplace Assistant - Evaluation (Agent SDK v2)");
  console.log("=".repeat(70));

  try {
    // Create agent using SDK v2
    const agent = await createWorkplaceAssistant();

    console.log(`\n✅ Agent created: ${agent.id}`);
    console.log(`   Model: ${agent.model}`);
    console.log(`   Name: ${agent.name}`);
    console.log("=".repeat(70));

    // Run evaluation
    const results = await runEvaluation(agent.name!);

    // Calculate and save results
    calculateAndSaveResults(results);
  } catch (error: any) {
    console.log(`\n❌ Evaluation failed: ${error.message}`);
    console.log(`Stack trace: ${error.stack}`);
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}
